sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java

```diff
@@ -0,0 +1,193 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.mqtt;
+
+import java.io.Serializable;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.activemq.broker.BrokerService;
+import org.apache.activemq.store.memory.MemoryPersistenceAdapter;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.testing.NeedsRunner;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.values.PCollection;
+import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
+import org.eclipse.paho.client.mqttv3.MqttCallback;
+import org.eclipse.paho.client.mqttv3.MqttClient;
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Tests of {@link MqttIO}.
+ */
+@RunWith(JUnit4.class)
+public class MqttIOTest implements Serializable {
+
+  private static transient BrokerService broker;
+
+  @BeforeClass
+  public static void startBroker() throws Exception {
+    broker = new BrokerService();
+    broker.setUseJmx(false);
+    broker.setPersistenceAdapter(new MemoryPersistenceAdapter());
+    broker.addConnector(new URI("mqtt://localhost:11883"));
+    broker.start();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testRead() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+
+    PCollection<byte[]> output = pipeline.apply(
+        MqttIO.read()
+            .withMqttConnectionConfiguration(
+                MqttIO.MqttConnectionConfiguration.create(
+                    "tcp://localhost:11883",
+                    "BEAM_PIPELINE",
+                    "READ_TOPIC"))
+          .withMaxNumRecords(10));
+
+    PAssert.thatSingleton(output.apply("Count", Count.<byte[]>globally()))
+        .isEqualTo(10L);
+    PAssert.that(output).satisfies(new SerializableFunction<Iterable<byte[]>, Void>() {
+
+      int count = 0;
+
+      @Override
+      public Void apply(Iterable<byte[]> input) {
+        for (byte[] element : input) {
+          String inputString = new String(element);
+          Assert.assertEquals("This is test " + count, inputString);
+          count++;
+        }
+        return null;
+      }
+    });
+
+    // produce messages on the broker in another thread
+    // This thread prevents to block the pipeline waiting for new messages
+    Thread thread = new Thread() {
+      public void run() {
+        try {
+          // gives time to the pipeline to start
```

Comment:
Explain why do we need to wait for the pipeline to start? Will MQTT drop messages if nobody's subscribed?

In general, I very strongly suggest to avoid timing-dependent tests. It always seems like you can choose a "large enough" delay, but I've again and again found that this is not the case, and environmental flakiness sooner or later causes the test to fail (maybe 1 in 1000 times - but at Beam's scale in terms of number of tests and number of test runs (e.g. pre-commits), pretty soon you have enough tests with this level of flakiness and pretty soon the overall flakiness grows unacceptable) and you have to rewrite the test. Is it possible to avoid this?